processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByRowProcessor.java

```diff
@@ -0,0 +1,508 @@
+                      }
+                  );
+                  filteredSequence.accumulate(grouper, accumulator);
+
+                  return new CloseableGrouperIterator<>(
+                      grouper,
+                      true,
+                      new Function<Grouper.Entry<GroupByMergingKey>, Row>()
+                      {
+                        @Override
+                        public Row apply(Grouper.Entry<GroupByMergingKey> entry)
+                        {
+                          Map<String, Object> theMap = Maps.newLinkedHashMap();
+
+                          // Add dimensions.
+                          for (int i = 0; i < entry.getKey().getDimensions().length; i++) {
+                            theMap.put(
+                                query.getDimensions().get(i).getOutputName(),
+                                Strings.emptyToNull(entry.getKey().getDimensions()[i])
+                            );
+                          }
+
+                          // Add aggregations.
+                          for (int i = 0; i < entry.getValues().length; i++) {
+                            theMap.put(query.getAggregatorSpecs().get(i).getName(), entry.getValues()[i]);
+                          }
+
+                          return new MapBasedRow(
+                              query.getGranularity().toDateTime(entry.getKey().getTimestamp()),
+                              theMap
+                          );
+                        }
+                      }
+                  );
+                }
+
+                @Override
+                public void cleanup(CloseableGrouperIterator<GroupByMergingKey, Row> iterFromMake)
+                {
+                  iterFromMake.close();
+                }
+              }
+          ),
+          new Closeable()
+          {
+            @Override
+            public void close() throws IOException
+            {
+              mergeBufferHolder.close();
+              CloseQuietly.close(temporaryStorage);
+            }
+          }
+      );
+    }
+    catch (Exception e) {
+      // Exception caught while creating the sequence; release resources.
+      mergeBufferHolder.close();
+      CloseQuietly.close(temporaryStorage);
+      throw e;
+    }
+  }
+
+  private static class RowBasedColumnSelectorFactory implements ColumnSelectorFactory
+  {
+    private Row row = null;
+
+    public void setRow(Row row)
+    {
+      this.row = row;
+    }
+
+    @Override
+    public DimensionSelector makeDimensionSelector(
+        DimensionSpec dimensionSpec
+    )
+    {
+      return dimensionSpec.decorate(makeDimensionSelectorUndecorated(dimensionSpec));
+    }
+
+    private DimensionSelector makeDimensionSelectorUndecorated(
+        DimensionSpec dimensionSpec
+    )
+    {
+      final String dimension = dimensionSpec.getDimension();
+      final ExtractionFn extractionFn = dimensionSpec.getExtractionFn();
+
+      return new DimensionSelector()
```

Comment:
This kind of dimensionSelector will make some aggregators not work properly, like Filtered. That's probably fine for now, but we should have a failing test and a follow up issue for fixing it.

Possible fixes might be:

- Making this DimensionSelector's ids be global instead of per-row (the assumption that the filtered agg depends on)
- Modify the DimensionSelector interface itself so it can inform callers if the column is really dictionary backed or not. Aggregators and filters would do different behavior on columns that aren't dictionary backed.

I think I like the second one.